// rsuite/cjs/internals/utils/date/index.js
'use client';
"use strict";

exports.__esModule = true;
var _exportNames = {
  addDays: true,
  addMonths: true,
  addYears: true,
  addSeconds: true,
  addMinutes: true,
  addHours: true,
  compareAsc: true,
  endOfDay: true,
  endOfISOWeek: true,
  endOfMonth: true,
  endOfWeek: true,
  format: true,
  getDate: true,
  getDay: true,
  getDaysInMonth: true,
  getHours: true,
  getMinutes: true,
  getMonth: true,
  getSeconds: true,
  getYear: true,
  isAfter: true,
  isBefore: true,
  isEqual: true,
  isSameDay: true,
  isSameMonth: true,
  isSameSecond: true,
  parse: true,
  parseISO: true,
  setDate: true,
  setHours: true,
  setMinutes: true,
  setMonth: true,
  setSeconds: true,
  setYear: true,
  startOfDay: true,
  startOfISOWeek: true,
  startOfMonth: true,
  startOfWeek: true,
  subDays: true,
  isMatch: true,
  isValid: true,
  set: true,
  differenceInCalendarMonths: true,
  isLastDayOfMonth: true,
  lastDayOfMonth: true,
  startOfToday: true,
  getWeekStartDates: true,
  getWeekKeys: true,
  reverseDateRangeOmitTime: true,
  omitHideDisabledProps: true,
  copyTime: true,
  disableTime: true,
  useDateMode: true,
  DateMode: true,
  extractTimeFormat: true,
  calendarOnlyProps: true
};
exports.useDateMode = exports.subDays = exports.startOfWeek = exports.startOfToday = exports.startOfMonth = exports.startOfISOWeek = exports.startOfDay = exports.setYear = exports.setSeconds = exports.setMonth = exports.setMinutes = exports.setHours = exports.setDate = exports.set = exports.reverseDateRangeOmitTime = exports.parseISO = exports.parse = exports.omitHideDisabledProps = exports.lastDayOfMonth = exports.isValid = exports.isSameSecond = exports.isSameMonth = exports.isSameDay = exports.isMatch = exports.isLastDayOfMonth = exports.isEqual = exports.isBefore = exports.isAfter = exports.getYear = exports.getWeekStartDates = exports.getWeekKeys = exports.getSeconds = exports.getMonth = exports.getMinutes = exports.getHours = exports.getDaysInMonth = exports.getDay = exports.getDate = exports.format = exports.extractTimeFormat = exports.endOfWeek = exports.endOfMonth = exports.endOfISOWeek = exports.endOfDay = exports.disableTime = exports.differenceInCalendarMonths = exports.copyTime = exports.compareAsc = exports.calendarOnlyProps = exports.addYears = exports.addSeconds = exports.addMonths = exports.addMinutes = exports.addHours = exports.addDays = exports.DateMode = void 0;
var _addDays = require("date-fns/addDays");
exports.addDays = _addDays.addDays;
var _addMonths = require("date-fns/addMonths");
exports.addMonths = _addMonths.addMonths;
var _addYears = require("date-fns/addYears");
exports.addYears = _addYears.addYears;
var _addSeconds = require("date-fns/addSeconds");
exports.addSeconds = _addSeconds.addSeconds;
var _addMinutes = require("date-fns/addMinutes");
exports.addMinutes = _addMinutes.addMinutes;
var _addHours = require("date-fns/addHours");
exports.addHours = _addHours.addHours;
var _compareAsc = require("date-fns/compareAsc");
exports.compareAsc = _compareAsc.compareAsc;
var _endOfDay = require("date-fns/endOfDay");
exports.endOfDay = _endOfDay.endOfDay;
var _endOfISOWeek = require("date-fns/endOfISOWeek");
exports.endOfISOWeek = _endOfISOWeek.endOfISOWeek;
var _endOfMonth = require("date-fns/endOfMonth");
exports.endOfMonth = _endOfMonth.endOfMonth;
var _endOfWeek = require("date-fns/endOfWeek");
exports.endOfWeek = _endOfWeek.endOfWeek;
var _format = require("date-fns/format");
exports.format = _format.format;
var _getDate = require("date-fns/getDate");
exports.getDate = _getDate.getDate;
var _getDay = require("date-fns/getDay");
exports.getDay = _getDay.getDay;
var _getDaysInMonth = require("date-fns/getDaysInMonth");
exports.getDaysInMonth = _getDaysInMonth.getDaysInMonth;
var _getHours = require("date-fns/getHours");
exports.getHours = _getHours.getHours;
var _getMinutes = require("date-fns/getMinutes");
exports.getMinutes = _getMinutes.getMinutes;
var _getMonth = require("date-fns/getMonth");
exports.getMonth = _getMonth.getMonth;
var _getSeconds = require("date-fns/getSeconds");
exports.getSeconds = _getSeconds.getSeconds;
var _getYear = require("date-fns/getYear");
exports.getYear = _getYear.getYear;
var _isAfter = require("date-fns/isAfter");
exports.isAfter = _isAfter.isAfter;
var _isBefore = require("date-fns/isBefore");
exports.isBefore = _isBefore.isBefore;
var _isEqual = require("date-fns/isEqual");
exports.isEqual = _isEqual.isEqual;
var _isSameDay = require("date-fns/isSameDay");
exports.isSameDay = _isSameDay.isSameDay;
var _isSameMonth = require("date-fns/isSameMonth");
exports.isSameMonth = _isSameMonth.isSameMonth;
var _isSameSecond = require("date-fns/isSameSecond");
exports.isSameSecond = _isSameSecond.isSameSecond;
var _parse = require("date-fns/parse");
exports.parse = _parse.parse;
var _parseISO = require("date-fns/parseISO");
exports.parseISO = _parseISO.parseISO;
var _setDate = require("date-fns/setDate");
exports.setDate = _setDate.setDate;
var _setHours = require("date-fns/setHours");
exports.setHours = _setHours.setHours;
var _setMinutes = require("date-fns/setMinutes");
exports.setMinutes = _setMinutes.setMinutes;
var _setMonth = require("date-fns/setMonth");
exports.setMonth = _setMonth.setMonth;
var _setSeconds = require("date-fns/setSeconds");
exports.setSeconds = _setSeconds.setSeconds;
var _setYear = require("date-fns/setYear");
exports.setYear = _setYear.setYear;
var _startOfDay = require("date-fns/startOfDay");
exports.startOfDay = _startOfDay.startOfDay;
var _startOfISOWeek = require("date-fns/startOfISOWeek");
exports.startOfISOWeek = _startOfISOWeek.startOfISOWeek;
var _startOfMonth = require("date-fns/startOfMonth");
exports.startOfMonth = _startOfMonth.startOfMonth;
var _startOfWeek = require("date-fns/startOfWeek");
exports.startOfWeek = _startOfWeek.startOfWeek;
var _subDays = require("date-fns/subDays");
exports.subDays = _subDays.subDays;
var _isMatch = require("date-fns/isMatch");
exports.isMatch = _isMatch.isMatch;
var _isValid = require("date-fns/isValid");
exports.isValid = _isValid.isValid;
var _set = require("date-fns/set");
exports.set = _set.set;
var _differenceInCalendarMonths = require("date-fns/differenceInCalendarMonths");
exports.differenceInCalendarMonths = _differenceInCalendarMonths.differenceInCalendarMonths;
var _isLastDayOfMonth = require("date-fns/isLastDayOfMonth");
exports.isLastDayOfMonth = _isLastDayOfMonth.isLastDayOfMonth;
var _lastDayOfMonth = require("date-fns/lastDayOfMonth");
exports.lastDayOfMonth = _lastDayOfMonth.lastDayOfMonth;
var _startOfToday = require("date-fns/startOfToday");
exports.startOfToday = _startOfToday.startOfToday;
var _getWeekStartDates = require("./getWeekStartDates");
exports.getWeekStartDates = _getWeekStartDates.getWeekStartDates;
var _getWeekKeys = require("./getWeekKeys");
exports.getWeekKeys = _getWeekKeys.getWeekKeys;
var _reverseDateRangeOmitTime = require("./reverseDateRangeOmitTime");
exports.reverseDateRangeOmitTime = _reverseDateRangeOmitTime.reverseDateRangeOmitTime;
var _omitHideDisabledProps = require("./omitHideDisabledProps");
exports.omitHideDisabledProps = _omitHideDisabledProps.omitHideDisabledProps;
var _copyTime = require("./copyTime");
exports.copyTime = _copyTime.copyTime;
var _disableTime = require("./disableTime");
exports.disableTime = _disableTime.disableTime;
var _useDateMode = require("./useDateMode");
exports.useDateMode = _useDateMode.useDateMode;
exports.DateMode = _useDateMode.DateMode;
var _extractTimeFormat = require("./extractTimeFormat");
exports.extractTimeFormat = _extractTimeFormat.extractTimeFormat;
var _formatCheck = require("./formatCheck");
Object.keys(_formatCheck).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _formatCheck[key]) return;
  exports[key] = _formatCheck[key];
});
var _types = require("./types");
exports.calendarOnlyProps = _types.calendarOnlyProps;
